signer/src/bitcoin/validation.rs

```diff
@@ -882,21 +907,67 @@ pub struct WithdrawalRequestReport {
     pub is_accepted: Option<bool>,
     /// The height of the bitcoin chain tip during the execution of the
     /// contract call that generated the withdrawal request.
-    pub block_height: u64,
+    pub bitcoin_block_height: u64,
 }
 
 impl WithdrawalRequestReport {
     /// Validate that the withdrawal request is okay given the report.
+    ///
+    /// See https://github.com/stacks-network/sbtc/issues/741 for the
+    /// validation rules for withdrawal requests.
     pub fn validate<F>(
         &self,
-        _: u64,
-        _: &F,
-        _: Amount,
-        _sbtc_limits: &SbtcLimits,
+        bitcoin_chain_tip_height: u64,
+        output_index: usize,
+        tx: &F,
+        tx_fee: Amount,
+        sbtc_limits: &SbtcLimits,
     ) -> WithdrawalValidationResult
     where
         F: FeeAssessment,
     {
+        match self.status {
+            WithdrawalRequestStatus::Confirmed => {}
+            WithdrawalRequestStatus::Unconfirmed => {
+                return WithdrawalValidationResult::TxNotOnBestChain
+            }
+            WithdrawalRequestStatus::Fulfilled(_) => {
+                return WithdrawalValidationResult::RequestFulfilled
+            }
+        }
+
+        match self.is_accepted {
+            Some(true) => (),
+            None => return WithdrawalValidationResult::NoVote,
+            Some(false) => return WithdrawalValidationResult::RequestRejected,
+        }
+
+        if self.amount > sbtc_limits.per_withdrawal_cap().to_sat() {
+            return WithdrawalValidationResult::AmountTooHigh;
+        }
+
+        if self.amount < WITHDRAWAL_DUST_LIMIT {
+            return WithdrawalValidationResult::AmountIsDust;
+        }
+
+        let block_wait = bitcoin_chain_tip_height.saturating_sub(self.bitcoin_block_height);
+        if block_wait < WITHDRAWAL_BLOCKS_WAIT {
+            return WithdrawalValidationResult::RequestNotFinal;
+        }
+
+        if block_wait > WITHDRAWAL_BLOCKS_EXPIRY {
+            return WithdrawalValidationResult::RequestExpired;
+        }
+
+        let Some(assessed_fee) = tx.assess_output_fee(output_index, tx_fee) else {
+            // If we hit this, then there is a programming error somewhere
+            return WithdrawalValidationResult::Unknown;
+        };
+
+        if assessed_fee.to_sat() > self.max_fee.min(self.amount) {
```

Comment:
Why do we want the min here? We are not deducing the fees from the amount for withdrawals as they are already locked (and will be burned) on the stacks side

Reply:
We don't want the min. Sorry, this was just a sloppy copy paste. I'll add a test case for this error too.

Reply:
Updated in https://github.com/stacks-network/sbtc/pull/1352/commits/3c088943b7d4cd355cb2dca1c90f16d8a30f4905